File: com-mini-core/src/main/java/com/mini/core/holder/jdbc/Join.java
package com.mini.core.holder.jdbc;

import com.mini.core.jdbc.builder.SQLBuilder;

import java.lang.annotation.*;

@Documented
@Target({ElementType.TYPE})
@Repeatable(Join.Joins.class)
@Retention(RetentionPolicy.RUNTIME)
public @interface Join {
	String column();
	
	JoinType type() default JoinType.INNER;
	
	enum JoinType {
		INNER {
			public void execute(SQLBuilder builder, String format, Object... args) {
				builder.join(format, args);
			}
		},
		LEFT {
			public void execute(SQLBuilder builder, String format, Object... args) {
				builder.leftJoin(format, args);
			}
		},
		RIGHT {
			public void execute(SQLBuilder builder, String format, Object... args) {
				builder.rightJoin(format, args);
			}
		},
		OUTER {
			public void execute(SQLBuilder builder, String format, Object... args) {
				builder.outerJoin(format, args);
			}
		};
		
		public abstract void execute(SQLBuilder builder, String format, Object... args);
	}
	
	@Documented
	@Target({ElementType.TYPE})
	@Retention(RetentionPolicy.RUNTIME)
	@interface Joins {
		Join[] value() default {};
	}
	
}

File: mini-plugin/src/main/kotlin/com/mini/core/jdbc/suppor/Join.kt
@file:Suppress("unused")

package com.mini.core.jdbc.suppor


import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.CLASS

@MustBeDocumented
@kotlin.annotation.Repeatable
@kotlin.annotation.Target(CLASS)
@kotlin.annotation.Retention(RUNTIME)
annotation class Join(val type: JoinType = JoinType.DEF, val value: String) {
    enum class JoinType {
        DEF,
        LEFT,
        INNER,
        RIGHT,
        OUTER;
    }

    @MustBeDocumented
    @kotlin.annotation.Target(CLASS)
    @kotlin.annotation.Retention(RUNTIME)
    annotation class List(vararg val value: Join = [])
}